flutter/lib/src/sentry_asset_bundle.dart

```diff
@@ -0,0 +1,176 @@
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+import 'package:flutter/material.dart';
+import 'package:sentry/sentry.dart';
+
+typedef _Parser<T> = Future<T> Function(String value);
+
+/// An [AssetBundle] which creates automatic performance traces for loading
+/// assets.
+///
+/// You can wrap other [AssetBundle]s in it:
+/// ```dart
+/// SentryAssetBundle(bundle: someOtherAssetBundle)
+/// ```
+/// If you're not providing any [AssetBundle], it falls back to the [rootBundle].
+///
+/// If you want to use the [SentryAssetBundle] by default you can achieve this
+/// with the following code:
+/// ```dart
+/// DefaultAssetBundle(
+///   bundle: SentryAssetBundle(),
+///   child: MaterialApp(
+///     home: MyScaffold(),
+///   ),
+/// );
+/// ```
+/// [Image.asset], for example, will then use [SentryAssetBundle].
+///
+/// The `enableStructureDataTracing` setting is an experimental feature.
+/// Use at your own risk.
+class SentryAssetBundle implements AssetBundle {
+  SentryAssetBundle({
+    Hub? hub,
+    AssetBundle? bundle,
+    bool enableStructureDataTracing = false,
+  })  : _hub = hub ?? HubAdapter(),
+        _bundle = bundle ?? rootBundle,
+        _enableStructureDataTracing = enableStructureDataTracing;
+
+  final Hub _hub;
+  final AssetBundle _bundle;
+  final bool _enableStructureDataTracing;
+
+  @override
+  Future<ByteData> load(String key) async {
+    final span = _hub.getSpan()?.startChild(
+          'file.read',
+          description: 'AssetBundle.load(key=$key)',
+        );
+
+    ByteData? data;
+    try {
+      data = await _bundle.load(key);
+      span?.status = SpanStatus.ok();
+    } catch (exception) {
+      span?.throwable = exception;
+      span?.status = SpanStatus.internalError();
+      rethrow;
+    } finally {
+      await span?.finish();
+    }
+    return data;
+  }
+
+  @override
+  Future<T> loadStructuredData<T>(String key, _Parser<T> parser) {
+    if (_enableStructureDataTracing) {
+      return _loadStructuredDataWithTracing(key, parser);
+    }
+    return _bundle.loadStructuredData(key, parser);
+  }
+
+  Future<T> _loadStructuredDataWithTracing<T>(
```

Comment:
@brustolin ideas on how to make it work without this hacky `runZonedGuarded`?

Reply:
If you get this to work I would love to get an explanation

Reply:
I believe it's not possible.

This is a part of flutter's `AssetBundle.loadStruturedData` function: 

```dart
    ...
    Completer<T>? completer;
    Future<T>? result;
    loadString(key, cache: false).then<T>(parser).then<void>((T value) {
      result = SynchronousFuture<T>(value);
      _structuredDataCache[key] = result!;
      if (completer != null) {
        // We already returned from the loadStructuredData function, which means
        // we are in the asynchronous mode. Pass the value to the completer. The
        // completer's future is what we returned.
        completer.complete(value);
      }
    });
    ...
```

As you can see `loadString` is not awaited, so, it's asynchronous execution is done in a different scope, not handled by a try that could be used instead of the `runZoneGuarded`.

In my understanding this is a dart bad approach.

But, that being said, we are changing the behavior of Flutter here. We are allowing and exception that should be thrown not matter what (by dart's design) to be handled.

Reply:
> But, that being said, we are changing the behavior of Flutter here. We are allowing and exception that should be thrown not matter what (by dart's design) to be handled.

That's why I've put it behind a flag which is disabled by default.

I'm okay with removing it and with keeping it. It's your call.

Reply:
@bruno-garcia opinions? I'd rather not do it at all (documenting this specific use case - telling people to do custom instrumentation) or keep this flag false by default, it still could lead to bugs anyway when people turn it on.

Reply:
There's a lot to unpack here and I have no context, is there a summary of what's going on?

Reply:
I had a call with @bruno-garcia and agreed to keep the flag disabled by default for now, till we get some feedback during the beta tests, if all is good, we flip during the next major.
@ueman just to confirm, the behavior is still the same when using it's own `runZoneGuarded`, right? the question is if the `rethrow` exception works correctly.

Reply:
never mind, the test answers the question :D thanks.